File: examples/android/FgVadDemo/app/src/main/java/io/fengur/fgvaddemo/WavWriter.kt
package io.fengur.fgvaddemo

import java.io.File
import java.io.RandomAccessFile

/**
 * 流式写 16 kHz mono PCM-16 WAV。
 * 用法：
 *   val w = WavWriter(file)
 *   w.append(samples, count)  // 多次
 *   w.finalize()              // 必调
 */
class WavWriter(private val file: File) {

    private val raf: RandomAccessFile
    private var samplesWritten: Int = 0

    init {
        file.parentFile?.mkdirs()
        if (file.exists()) file.delete()
        raf = RandomAccessFile(file, "rw")
        raf.write(placeholderHeader())   // 占位 header，finalize 时回填
    }

    fun append(samples: ShortArray, count: Int) {
        val n = count.coerceAtMost(samples.size)
        val bytes = ByteArray(n * 2)
        for (i in 0 until n) {
            val s = samples[i].toInt()
            bytes[2 * i] = (s and 0xff).toByte()
            bytes[2 * i + 1] = ((s shr 8) and 0xff).toByte()
        }
        raf.write(bytes)
        samplesWritten += n
    }

    fun finalize() {
        val dataBytes = samplesWritten * 2
        val riffSize = 36 + dataBytes
        // RIFF size at offset 4
        raf.seek(4); raf.write(le32(riffSize))
        // data size at offset 40
        raf.seek(40); raf.write(le32(dataBytes))
        raf.close()
    }

    private fun placeholderHeader(): ByteArray {
        val out = java.io.ByteArrayOutputStream(44)
        out.write("RIFF".toByteArray(Charsets.US_ASCII))
        out.write(le32(0))                         // riff size, finalize 回填
        out.write("WAVE".toByteArray(Charsets.US_ASCII))
        out.write("fmt ".toByteArray(Charsets.US_ASCII))
        out.write(le32(16))                        // fmt chunk size
        out.write(le16(1))                         // audio format = PCM
        out.write(le16(1))                         // channels
        out.write(le32(16000))                     // sample rate
        out.write(le32(16000 * 2))                 // byte rate
        out.write(le16(2))                         // block align
        out.write(le16(16))                        // bits per sample
        out.write("data".toByteArray(Charsets.US_ASCII))
        out.write(le32(0))                         // data size, finalize 回填
        return out.toByteArray()
    }

    private fun le32(v: Int): ByteArray = byteArrayOf(
        (v and 0xff).toByte(),
        ((v shr 8) and 0xff).toByte(),
        ((v shr 16) and 0xff).toByte(),
        ((v shr 24) and 0xff).toByte(),
    )
    private fun le16(v: Int): ByteArray = byteArrayOf(
        (v and 0xff).toByte(),
        ((v shr 8) and 0xff).toByte(),
    )
}

File: examples/ios/FgVadDemo/FgVadDemo/WavWriter.swift
import Foundation

/// 流式写 16 kHz mono PCM-16 WAV。
/// 用法：
///   let w = try WavWriter(url: url)
///   w.append(samples: [Int16])  // 多次
///   try w.finalize()             // 必调，否则 header size 字段是 0，文件不可读
final class WavWriter {

    private let handle: FileHandle
    private var samplesWritten: UInt32 = 0

    init(url: URL) throws {
        let fm = FileManager.default
        try fm.createDirectory(at: url.deletingLastPathComponent(),
                               withIntermediateDirectories: true)
        if fm.fileExists(atPath: url.path) {
            try fm.removeItem(at: url)
        }
        fm.createFile(atPath: url.path, contents: nil)
        self.handle = try FileHandle(forWritingTo: url)
        try handle.write(contentsOf: WavWriter.placeholderHeader())
    }

    func append(samples: UnsafePointer<Int16>, count: Int) throws {
        let bytes = count * MemoryLayout<Int16>.size
        let data = Data(bytes: samples, count: bytes)
        try handle.write(contentsOf: data)
        samplesWritten += UInt32(count)
    }

    func append(samples: [Int16]) throws {
        try samples.withUnsafeBufferPointer { buf in
            guard let base = buf.baseAddress else { return }
            try append(samples: base, count: buf.count)
        }
    }

    func finalize() throws {
        // 回填 RIFF chunk size + data chunk size
        let dataBytes = samplesWritten * 2
        let riffSize = 36 + dataBytes
        try handle.seek(toOffset: 4)
        try handle.write(contentsOf: WavWriter.le32(riffSize))
        try handle.seek(toOffset: 40)
        try handle.write(contentsOf: WavWriter.le32(dataBytes))
        try handle.close()
    }

    private static func placeholderHeader() -> Data {
        var d = Data()
        d.append("RIFF".data(using: .ascii)!)
        d.append(le32(0))                       // riff size, finalize 时回填
        d.append("WAVE".data(using: .ascii)!)
        d.append("fmt ".data(using: .ascii)!)
        d.append(le32(16))                      // fmt chunk size
        d.append(le16(1))                       // audio format = PCM
        d.append(le16(1))                       // channels = 1
        d.append(le32(16000))                   // sample rate
        d.append(le32(16000 * 2))               // byte rate = sample rate * block align
        d.append(le16(2))                       // block align = channels * bits/8
        d.append(le16(16))                      // bits per sample
        d.append("data".data(using: .ascii)!)
        d.append(le32(0))                       // data size, finalize 时回填
        return d
    }

    private static func le32(_ v: UInt32) -> Data {
        var x = v.littleEndian
        return Data(bytes: &x, count: 4)
    }
    private static func le16(_ v: UInt16) -> Data {
        var x = v.littleEndian
        return Data(bytes: &x, count: 2)
    }
    private static func le32(_ v: Int) -> Data { le32(UInt32(v)) }
    private static func le16(_ v: Int) -> Data { le16(UInt16(v)) }
}
